networkx/algorithms/coloring/edge_coloring.py

```diff
@@ -0,0 +1,147 @@
+import networkx as nx
+
+__all__ = ["edge_coloring"]
+
+
+@nx.utils.not_implemented_for("directed")
+@nx.utils.not_implemented_for("multigraph")
+@nx._dispatchable
+def edge_coloring(G):
+    """This function performs edge coloring on a given graph G.
+
+    Parameters
+    ----------
+    G: NetworkX graph object representing the input graph.
+
+    Returns
+    -------
+    coloring: A dictionary where keys are tuples representing edges of the graph,
+                and values are integers representing colors assigned to the edges.
+
+    Edge coloring is the assignment of colors to the edges of a graph in such a way
+    that no two adjacent edges share the same color.
+
+    References
+    ----------
+    This implementation follows the proof of the lemma given in Zdenek Dvorak's lecture notes
+    titled Graph coloring: Heawood and Brooks theorems, edge coloring which states
+    Let G be a graph of maximum degree at most d with edges properly colored using
+    d + 1 colors. Let u and v be distinct non-adjacent vertices of G.
+    Then there exists a proper edge coloring using d + 1 colors
+    such that the same color missing at both u and v.
+    https://iuuk.mff.cuni.cz/~rakdver/kgii/lesson20-6.pdf
+    """
+
+    # function to color an edge
+    def color_edge(edge, color):
+        u, v = edge
+        used_colors[u][color] = v
+        used_colors[v][color] = u
+        coloring[(u, v)] = color
+        coloring[(v, u)] = color
+
+    G = G.copy()
+    degrees = dict(G.degree())
+    delta = max(degrees.values())
+    colors = set(range(delta + 1))
```

Comment:
Maybe use the range for colors instead of the set.  
And make `u_colors` be the dict `used_colors[u]` and only use the `keys()` when doing set operations.
(similar comment to the bipartite version)

Reply:
We are keeping the colors as set to make it easier to find available colors using set difference.